### ISSUE TITLE
Failure and inconsistent state when job takes too long to start

### ISSUE THREAD

**Issue by rkdarst** [jupyterhub/batchspawner#88]
I think that batchspawner may get into an inconsistent state if the `.start()` method never returns.  I don't have a full analysis yet, but I'm pasting logs here and will try to interpret tomorrow.  I've noticed something like this from months ago, when I first set up jupyterhub, too, but didn't go so far as to determine the cause and I'm not sure it's exactly the same thing.

My first hypothesis is that `.start()` times out on the jupyterhub side, leaving things in an inconsistent state.  That would be this:

```
[E 2018-05-24 00:03:44.158 JupyterHub gen:940] Exception in Future <Task finished coro=<BaseHandler.spawn_single_user.<locals>.finish_user_spawn() done, defined at /share/apps/jupyterhub/live/jupyterhub/jupyter\
hub/handlers/base.py:617> exception=TimeoutError('Timeout',)> after timeout
Traceback (most recent call last):
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/handlers/base.py", line 624, in finish_user_spawn
await spawn_future
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/user.py", line 484, in spawn
raise e
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/user.py", line 404, in spawn
url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
tornado.util.TimeoutError: Timeout
[E 2018-05-24 00:03:44.172 JupyterHub gen:940] Exception in Future <Task finished coro=<BaseHandler.spawn_single_user.<locals>.finish_user_spawn() done, defined at /share/apps/jupyterhub/live/jupyterhub/jupyter\
hub/handlers/base.py:617> exception=TimeoutError('Timeout',)> after timeout
Traceback (most recent call last):
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936, in error_callback
future.result()
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/handlers/base.py", line 624, in finish_user_spawn
await spawn_future
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/user.py", line 484, in spawn
raise e
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/user.py", line 404, in spawn
url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
tornado.util.TimeoutError: Timeout
tornado.util.TimeoutError: Timeout
[E 2018-05-24 00:03:44.173 JupyterHub gen:940] Exception in Future <Task finished coro=<BaseHandler.spawn_single_user.<locals>.finish_user_spawn() done, defined at /share/apps/jupyterhub/live/jupyterhub/jupyter\
hub/handlers/base.py:617> exception=TimeoutError('Timeout',)> after timeout
Traceback (most recent call last):
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/handlers/base.py", line 624, in finish_user_spawn
await spawn_future
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/user.py", line 484, in spawn
raise e
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/user.py", line 404, in spawn
url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
tornado.util.TimeoutError: Timeout
[W 2018-05-24 00:03:44.175 JupyterHub users:439] Stream closed while handling /hub/api/users/username/server/progress
[W 2018-05-24 00:03:44.175 JupyterHub users:439] Stream closed while handling /hub/api/users/username/server/progress
[W 2018-05-24 00:03:44.175 JupyterHub users:439] Stream closed while handling /hub/api/users/username/server/progress
[I 2018-05-24 00:03:44.177 JupyterHub log:158] 200 GET /hub/api/users/username/server/progress (username@::ffff:127.0.0.1) 159088.94ms
[I 2018-05-24 00:03:44.178 JupyterHub log:158] 200 GET /hub/api/users/username/server/progress (username@::ffff:127.0.0.1) 125638.94ms
[I 2018-05-24 00:03:44.178 JupyterHub log:158] 200 GET /hub/api/users/username/server/progress (username@::ffff:127.0.0.1) 93877.60ms
[I 2018-05-24 00:03:44.180 JupyterHub log:158] 200 GET /hub/api/users/username/server/progress (username@::ffff:127.0.0.1) 28713.99ms
[W 2018-05-24 00:03:44.411 JupyterHub batchspawner:357] Job  neither pending nor running.
[E 2018-05-24 00:03:44.412 JupyterHub gen:940] Exception in Future <Future finished exception=AssertionError()> after timeout
Traceback (most recent call last):
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/batchspawner/batchspawner/batchspawner.py", line 358, in start
assert self.state_ispending()
AssertionError
```

Then, when you try to go to hub/user/username/ in the future, you get a 500 server error, which I presume is this:

```
[E 2018-05-24 00:07:39.081 JupyterHub base:939] Preventing implicit spawn for username because lastspawn failed: Timeout
[E 2018-05-24 00:07:39.081 JupyterHub web:1621] Uncaught exception GET /hub/user/username/ (::ffff:1
                                                                                            27.0.0.1)
HTTPServerRequest(protocol='http', host='xxx', method='GET', uri='/hub/user/username/', version='HTTP/1.1', remote_ip='::ffff:127.0.0.1')
Traceback (most recent call last):
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/web.py", line 1543,in _execute
result = yield result
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/handlers/base.py", line 941, in get
raise copy.copy(exc).with_traceback(exc.__traceback__)
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/miniconda/lib/python3.6/site-packages/tornado/gen.py", line 936,in error_callback
future.result()
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/handlers/base.py", line 624, in finish_user_spawn
await spawn_future
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/user.py", line 484, in spawn
raise e
File "/share/apps/jupyterhub/live/jupyterhub/jupyterhub/user.py", line 404, in spawn
url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
tornado.util.TimeoutError: Timeout
[E 2018-05-24 00:07:39.089 JupyterHub log:150] {
"X-Forwarded-Proto": "http",
"X-Forwarded-Port": "80",
"Connection": "close",
"X-Forwarded-Server": "xxx",
"X-Forwarded-Host": "xxx",
"X-Forwarded-For": "...,::ffff:127.0.0.1",
"Upgrade-Insecure-Requests": "1",
"Dnt": "1",
"Cookie": "...",
"Accept-Encoding": "gzip, deflate, br",
"Accept-Language": "en-US,en;q=0.5",
"Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
"User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:59.0) Gecko/20100101 Firefox/59.0",
"Host": "xxx"
}
[E 2018-05-24 00:07:39.089 JupyterHub log:158] 500 GET /hub/user/username/ (username@::ffff:127.0.0.1) 16.39ms
```

Note that /hub/spawn *does* work and I can spawn a server again.

I will try to understand more tomorrow...

**Comment by rkdarst** [comment]
I just saw this referenced so I should clarify...

In the end, it wasn't on inconsistest state, but what I suggested fixed in https://github.com/jupyterhub/jupyterhub/pull/1915.  Basically, when I went to the same page again, I saw a 500 error message which made me think that something went wrong in the server, but really it was re-reporting the previous error.  Once I understood, I could go to the (obvious) link and try to respawn again, which would work if (if respawning didn't fail again...).

So I think I can close this now, things are "clear enough" for me... unless someone else knows another problem.
